File: csrc/minimumSound.c
#include <windows.h>
#include <Mmsystem.h>
#include <Mmreg.h>
#include <math.h>

#define WIDTH 640
#define HEIGHT 480

#define SOUND_SAMPLE_RATE		44100
#define SOUND_LENGTH_IN_SECOND	60
#define SOUND_NUM_CHANNELS		2
#define SOUND_NUM_SAMPLES		(SOUND_SAMPLE_RATE*SOUND_LENGTH_IN_SECOND)

typedef float sample_type;

sample_type	samples[SOUND_NUM_SAMPLES * SOUND_NUM_CHANNELS];

WAVEFORMATEX wave_format =
{
	WAVE_FORMAT_IEEE_FLOAT,
	SOUND_NUM_CHANNELS,							//nChannels
	SOUND_SAMPLE_RATE,							//nSamplesPerSec,
	SOUND_SAMPLE_RATE*sizeof(sample_type)*
	SOUND_NUM_CHANNELS,							//nAvgBytesPerSec
	sizeof(sample_type)*SOUND_NUM_CHANNELS,		//nBlockAlign
	sizeof(sample_type)*8,						//wBitsPerSample
	0
};

WAVEHDR wave_hdr =
{
	(LPSTR)samples,
	sizeof(samples),
	0,
	0,
	0,
	0,
	0,
	0
};

void WinMainCRTStartup(void) {
  HWND hWnd =	CreateWindowA("STATIC", 0,
      WS_POPUP|WS_VISIBLE, 0, 0,
      WIDTH, HEIGHT, NULL, NULL, NULL, NULL);

#define PI	3.1415926f

  for(int i = 0; i < SOUND_NUM_SAMPLES; ++i) {
    float t = (float)i/SOUND_SAMPLE_RATE;
    float v = sinf(2.0f * PI * t * 400.0f) * 0.25f;
    samples[i*2] = v;
    samples[i*2 + 1] = v;
  }

  HWAVEOUT h_wave_out;
	waveOutOpen(&h_wave_out, WAVE_MAPPER, &wave_format, (DWORD_PTR)hWnd, 0, CALLBACK_WINDOW);
	waveOutPrepareHeader(h_wave_out, &wave_hdr, sizeof(wave_hdr));
	waveOutWrite(h_wave_out, &wave_hdr, sizeof(wave_hdr));

	MSG msg;
loop:
	PeekMessage(&msg, 0, 0, 0, PM_REMOVE);
	if(GetAsyncKeyState(VK_ESCAPE) || msg.message == MM_WOM_DONE)
		ExitProcess(0);
  Sleep(256);
	goto loop;
}

File: src/minimumSound.nim
import winlean4k
import math

type SampleType = float32

const
  soundSampleRate     = 44100
  soundLengthInSecond = 60
  soundNumChannels    = 2
  soundNumSamples     = soundSampleRate * soundLengthInSecond

let
  waveFormat = WAVEFORMATEX(
    wFormatTag:       WAVE_FORMAT_IEEE_FLOAT,
    nChannels:        soundNumChannels,
    nSamplesPerSec:   soundSampleRate,
    nAvgBytesPerSec:  soundSampleRate*sizeof(SampleType)*soundNumChannels,
    nBlockAlign:      sizeof(SampleType)*soundNumChannels,
    wBitsPerSample:   sizeof(SampleType)*8,
    cbSize:           0
  )

var samples: array[soundNumSamples * soundNumChannels, SampleType]

var
  waveHdr = WAVEHDR(
    lpData:           nil,
    dwBufferLength:   sizeof(samples).uint32,
    dwBytesRecorded:  0,
    dwUser:           0,
    dwFlags:          0,
    dwLoops:          0,
    lpNext:           nil,
    reserved:         0
  )

template checkWaveOutCall(call: typed): untyped =
  when defined(danger):
    discard call
  else:
    let r = call
    if r != MMSYSERR_NOERROR:
      var text: array[MAXERRORLENGTH, Utf16Char]
      if waveOutGetErrorTextW(r, addr text[0], MAXERRORLENGTH.UINT) != MMSYSERR_NOERROR:
        echo cast[WideCString](addr text[0])
      else:
        quit "Failed to call waveOutGetErrorText"

proc WinMainCRTStartup() {.exportc.} =
  let hWnd = CreateWindowA(
    "STATIC".cstring, nil,
    WS_POPUP or WS_VISIBLE, 0, 0,
    640, 480, nil, nil, nil, nil)

  for i in 0..<soundNumSamples:
    let
      t: float32 = float32(i)/soundSampleRate
      v: float32 = sin(2.0'f * float32(PI) * t * 400.0'f) * 0.25'f
    samples[i * 2] = v
    samples[i * 2 + 1] = v

  var h_wave_out: HWAVEOUT
  checkWaveOutCall(
                   waveOutOpen(
                               addr h_wave_out,
                               WAVE_MAPPER,
                               unsafeAddr waveFormat,
                               cast[DWORD_PTR](hWnd),
                               0.DWORD_PTR,
                               CALLBACK_WINDOW.DWORD))

  wave_hdr.lpData = cast[cstring](addr samples[0])
  checkWaveOutCall(
                   waveOutPrepareHeader(
                                        h_wave_out,
                                        addr waveHdr,
                                        sizeof(waveHdr).UINT))
  checkWaveOutCall(waveOutWrite(h_wave_out, addr waveHdr, sizeof(waveHdr).UINT))

  var msg {.noinit.}: MSG

  while true:
    discard PeekMessageA(addr msg, nil, 0, 0, PM_REMOVE)
    if GetAsyncKeyState(VK_ESCAPE) != 0 or msg.message == MM_WOM_DONE:
      ExitProcess(0)
    Sleep(256)

when not defined(danger):
  WinMainCRTStartup()
